File: residential_controller.rb
class Column
    @@no_of_column = 0
    def initialize(_id, _amountOfElevators,_amountOfFloors)
        @column_ID= _id
        @column_status = 'Online'
        @column_elevatorList = []
        @column_callButtonList =[]
        createElevators _amountOfElevators, _amountOfFloors 
        createCallButtons _amountOfFloors
    end    
        
    def createCallButtons(_amountOfFloors)
        buttonFloor = 1
        callButtonID = 1
        floor = 1
            until floor == _amountOfFloors do
                if buttonFloor < _amountOfFloors
                    callButton = CallButton(callButtonID, buttonFloor, 'up')
                    callButtonList.append(callButton)
                end
                if buttonFloor > 1
                    callButton = CallButton(callButtonID, buttonFloor, 'down')
                    callButtonList.append(callButton)
                end
                callButtonID += 1
                buttonFloor += 1
            return callButtonList
        end
    end
    def createElevators( _amountOfElevators, _amountOfFloors)
        elevatorID = 1 
        until elevatorID == _amountOfElevators do 
            elevator = Elevator(elevatorID, _amountOfFloors)
            elevatorList.append(elevator)
            elevatorID += 1
        return elevatorList
        end
    end
    # Simulate when a user press a button outside the elevator
    def requestElevator(floor,direction)
        best = findElevator(floor,direction)
        best.floorRequestList.append(floor)
        best.move()
        #  elevator.operateDoors()
        return best
    end 
    # We use a score system depending on the current elevators state. Since the bestScore and the referenceGap are 
    # higher values than what could be possibly calculated, the first elevator will always become the default bestElevator, 
    # before being compared with to other elevators. If two elevators get the same score, the nearest one is prioritized.
    def findElevator(requestedFloor, requestedDirection)
        # bestElevator = type('', (), {})()
        bestScore = 5
        referenceGap = 10000000
        bestElevatorInformations = []
        for elevator in elevatorList do
            # The elevator is at my floor and going in the direction I want
            if (requestedFloor == elevator.currentFloor) & (elevator.status == 'stopped') &(requestedDirection == elevator.direction)
                bestElevatorInformations = checkIfElevatorIsBetter(1, elevator, bestScore, referenceGap, bestElevator,requestedFloor)
            
            # The elevator is lower than me, is coming 'Up' and I want to go 'Up'
            elsif (requestedFloor > elevator.currentFloor) & (elevator.direction == 'up') &(requestedDirection == elevator.direction)
                bestElevatorInformations = checkIfElevatorIsBetter(2, elevator,bestScore, referenceGap, bestElevator, requestedFloor)
            
            # The elevator is higher than me, is coming 'Down' and I want to go 'Down'
            elsif (requestedFloor < elevator.currentFloor) & (elevator.direction == 'down') &(requestedDirection == elevator.direction)
                bestElevatorInformations = checkIfElevatorIsBetter(2, elevator,bestScore, referenceGap, bestElevator, requestedFloor)
            
            # The elevator is idle
            elsif elevator.status == 'idle'
                bestElevatorInformations = checkIfElevatorIsBetter(3, elevator,bestScore, referenceGap, bestElevator, requestedFloor)
            
            else
                bestElevatorInformations = checkIfElevatorIsBetter(4, elevator, bestScore, referenceGap, bestElevator, requestedFloor)
            end
            bestElevator = bestElevatorInformations[0]
            bestScore = bestElevatorInformations[1]
            referenceGap = bestElevatorInformations[2]
        return bestElevator
        end
    end
    

    def checkIfElevatorIsBetter(scoreToCheck, newElevator, bestScore, referenceGap, bestElevator, floor)
        if scoreToCheck < bestScore
             bestScore= scoreToCheck
             bestElevator = newElevator
             referenceGap = abs(newElevator.currentFloor - floor)
        
        elsif bestScore == scoreToCheck
            gap = abs(newElevator.currentFloor - floor)
            if referenceGap > gap
                 bestElevator = newElevator
                 referenceGap = gap
            end
        end    
        
        return [bestElevator, bestScore, referenceGap]
    
    end
end
class FloorRequestButton
    @@no_of_floorrequestbutton = 0
    def initialize(_id, _floor)
        @floorrequestbutton_ID= _id
        @floorrequestbutton_status = 'OFF'
        @floorrequestbutton_floor = _floor
    end    
end     
class Elevator 
    @@no_of_elevator = 0
    def initialize(_id, _amountOfFloors)
        @elevator_ID= _id
        @elevator_status = 'idle'
        @elevator_currentFloor = 1
        @elevator_direction = None
        @elevator_door = Door(_id)
        @elevator_floorRequestButtonList =[]
        @elevator_floorButtonsList =[]
        @elevator_floorRequestList = []
        createFloorRequestButtons(_amountOfFloors)
    end        
    def createFloorRequestButtons(_amountOfFloors)
        buttonFloor = 1
        buttonID = 1
        
            until buttonID  == _amountOfFloors do 
                floorRequestButton = FloorRequestButton(FloorRequestButtonID, buttonFloor)
                floorRequestButtonList.append(floorRequestButton)
                buttonFloor += 1
            buttonID += 1
        end
    end

            
    
    # Simulate when a user press a button inside the elevator
    def requestFloor(floor)
        floorRequestList.append(floor)
        move
        #  operateDoors()
    end    
    def move
        while len(floorRequestList) != 0 
            destination = floorRequestList[0]
            status = 'moving'
            if currentFloor < destination
                direction = 'up'
                sortFloorList()
                while currentFloor < destination
                    currentFloor += 1
                    screenDisplay = currentFloor
                end
            elsif currentFloor > destination
                direction = 'down'
                sortFloorList()
                while currentFloor > destination
                    currentFloor -= 1
                    screenDisplay = currentFloor
                end
            end 
            status = 'stopped'
            floorRequestList.pop(0)
        end
        status = 'idle'
    end
    def sortFloorList()
        if direction == 'up'
            floorRequestList.sort()
        else
            floorRequestList.reverse()
        end
    end
end
class CallButton
    @@no_of_callbutton = 0
    def initialize(_id, _floor, _direction)
        @callbutton_ID= _id
        @callbutton_status = "OFF"
        @callbutton_floor = _floor
        @callbutton_direction = _direction
    end    
end
class Door
    @@no_of_door = 0
    def initialize(_id)
        @door_ID= _id
        @door_status = 'closed'
    end    
end
    

File: residential_controller.py
class Column:
    def __init__(self, _id, _amountOfFloors, _amountOfElevators):
        self.ID= _id
        self.status = []
        self.elevatorList = []
        self.callButtonList =[]
        self.requestList =[]
        self.createElevators(_amountOfElevators, _amountOfFloors)
        self.createCallButtons(_amountOfFloors)
    def createCallButtons(self,  _amountOfFloors):
        buttonFloor = 1
        callButtonID = 1
        for x in range(_amountOfFloors):
            if (buttonFloor < _amountOfFloors):
                callButton = CallButton(callButtonID, buttonFloor, 'up')
                self.callButtonList.append(callButton)
                callButtonID += 1
            if (buttonFloor > 1):
                callButton = CallButton(callButtonID, buttonFloor, 'down')
                self.callButtonList.append(callButton)
                callButtonID += 1
            buttonFloor+=1
        return self.callButtonList
    
    def createElevators(self, _amountOfElevators, _amountOfFloors):
        elevatorID = 1 
        for x in range(_amountOfElevators):
            elevator = Elevator(elevatorID, _amountOfFloors)
            self.elevatorList.append(elevator)
            elevatorID += 1
        return self.elevatorList
    
    # Simulate when a user press a button outside the elevator
    def requestElevator(self, floor,direction):
        self.requestList.append(floor)
        best = self.findElevator(floor,direction)
        best.floorRequestList.append(floor)
        best.move()
        #  elevator.operateDoors()
        return best
    
    # We use a score system depending on the current elevators state. Since the bestScore and the referenceGap are 
    # higher values than what could be possibly calculated, the first elevator will always become the default bestElevator, 
    # before being compared with to other elevators. If two elevators get the same score, the nearest one is prioritized.
    def findElevator(self, requestedFloor, requestedDirection):
        bestElevator = type('', (), {})()
        bestScore = 5
        referenceGap = 10000000
        bestElevatorInformations = []
        for elevator in self.elevatorList:
            # The elevator is at my floor and going in the direction I want
            if (requestedFloor == elevator.currentFloor) & (elevator.status == 'stopped') &(requestedDirection == elevator.direction):
                bestElevatorInformations = self.checkIfElevatorIsBetter(1, elevator, bestScore, referenceGap, bestElevator,requestedFloor)
            
            # The elevator is lower than me, is coming 'Up' and I want to go 'Up'
            elif (requestedFloor > elevator.currentFloor) & (elevator.direction == 'up') &(requestedDirection == elevator.direction):
                bestElevatorInformations = self.checkIfElevatorIsBetter(2, elevator,bestScore, referenceGap, bestElevator, requestedFloor)
            
            # The elevator is higher than me, is coming 'Down' and I want to go 'Down'
            elif (requestedFloor < elevator.currentFloor) & (elevator.direction == 'down') &(requestedDirection == elevator.direction):
                bestElevatorInformations = self.checkIfElevatorIsBetter(2, elevator,bestScore, referenceGap, bestElevator, requestedFloor)
            
            # The elevator is idle
            elif elevator.status == 'idle':
                bestElevatorInformations = self.checkIfElevatorIsBetter(3, elevator,bestScore, referenceGap, bestElevator, requestedFloor)
            
            else:
                bestElevatorInformations = self.checkIfElevatorIsBetter(4, elevator, bestScore, referenceGap, bestElevator, requestedFloor)
            
            bestElevator = bestElevatorInformations[0]
            bestScore = bestElevatorInformations[1]
            referenceGap = bestElevatorInformations[2]
        return bestElevator
    

    def checkIfElevatorIsBetter(self, scoreToCheck, newElevator, bestScore, referenceGap, bestElevator, floor):
        if scoreToCheck < bestScore:
             bestScore= scoreToCheck
             bestElevator = newElevator
             referenceGap = abs(newElevator.currentFloor - floor)
        
        elif bestScore == scoreToCheck:
            gap = abs(newElevator.currentFloor - floor)
            if referenceGap > gap:
                 bestElevator = newElevator
                 referenceGap = gap
            
        
        return [bestElevator, bestScore, referenceGap]
    

class FloorRequestButton: 
    def __init__(self, _id, _floor):
        self.ID = _id
        self.status = 'OFF'
        self.floor = _floor
    
class Elevator:
    def __init__(self, _id, _amountOfFloors):
        self.ID = _id
        self.status = 'idle'
        self.currentFloor = 1
        self.direction = None
        self.door = Door(_id)
        self.floorRequestButtonList =[]
        self.floorButtonsList =[]
        self.floorRequestList = []
        self.createFloorRequestButtons(_amountOfFloors)
    
    def createFloorRequestButtons(self, _amountOfFloors):
        buttonFloor = 1
        
        FloorRequestButtonID = 1
        for i in range(_amountOfFloors):
            floorRequestButton = FloorRequestButton(FloorRequestButtonID, buttonFloor)
            self.floorRequestButtonList.append(floorRequestButton)
            buttonFloor += 1
            FloorRequestButtonID += 1
            
    
    # Simulate when a user press a button inside the elevator
    def requestFloor(self, floor):
        self.floorRequestList.append(floor)
        self.move()
        #  self.operateDoors()
    
    def move(self):
        while (len(self.floorRequestList) != 0 ):
            destination = self.floorRequestList[0]
            self.status = 'moving'
            if self.currentFloor < destination:
                self.direction = 'up'
                self.sortFloorList()
                while self.currentFloor < destination:
                    self.currentFloor += 1
                    self.screenDisplay = self.currentFloor
                
            
            elif self.currentFloor > destination:
                self.direction = 'down'
                self.sortFloorList()
                while self.currentFloor > destination:
                    self.currentFloor -= 1
                    self.screenDisplay = self.currentFloor
                
            
            self.status = 'stopped'
            self.floorRequestList.pop(0)
        
        self.status = 'idle'
    
    def sortFloorList(self):
        if self.direction == 'up':
            self.floorRequestList.sort()
        else:
            self.floorRequestList.reverse()
        

class CallButton:
    def __init__(self, _id, _floor, _direction):
        self.ID = _id
        self.status = "OFF"
        self.floor = _floor
        self.direction = _direction
class Door:
    def __init__(self, _id):
        self.ID = _id
        self.status = 'closed'

    

    
    #  operateDoors()
    #      self.door.status = opened
    #      await 5 seconds
    #      if (this is not overweight)
    #          self.door.status = closing
    #          if no obstruction
    #              self.door.status = closed
    #          
    #          else
    #              self.operateDoors()
    #          
    #      
    #      else
    #          while (this is overweight)
    #              activate overweight alarm
    #          
    #          self.operateDoors()
    #      
    #  